### DISCUSSION TITLE
Should "__Nonexhaustive" enum variant be recommended for future proofing?

### DISCUSSION THREAD

**Discussion by kjetilkjeka** [rust-lang-nursery/api-guidelines#173] [API Guidelines]
The `__Nonexhaustive` variant pattern is useful for allowing `enum`s to grow post release. One example of use from the standard library is `std::io::ErrorKind`, and a constructed example follows this paragraph. Should this pattern be mentioned in the guidelines under future proofing?

``` rust    
pub enum Example {
    Variant1,
    Variant2,
    // Omitted variants
    VariantN,
    /// A marker variant that tells the compiler that users of this enum cannot match it exhaustively.
    #[doc(hidden)]
    __Nonexhaustive,
}
```

**Comment by Ixrec** [comment]
Note that there's an accepted RFC for a `#[non_exhaustive]` attribute for precisely this use case: https://github.com/rust-lang/rust/issues/44109 But once that's implemented, I'd agree that an API guideline for using it would be good.

**Comment by kjetilkjeka** [comment]
I was not aware of this RFC, it looks like the feature will be stabilized soon. I agree that it's a good idea to wait for this then.

**Comment by jonas-schievink** [comment]
`#[non_exhaustive]` is now stable.

**Comment by KodrAus** [comment]
I think we should definitely be recommending `#[non_exhaustive]` on structs and enums with public fields or variants.

**Comment by dtolnay** [comment]
For the nonexhaustive enums in Syn, `#[non_exhaustive]` is unusable until https://github.com/rust-lang/rust/issues/44109#issuecomment-521781237 gets implemented.
